keyboards/bastardkb/charybdis/3x5/keymaps/via/keymap.c

```diff
@@ -0,0 +1,349 @@
+/**
+ * Copyright 2021 Charly Delay <charly@codesink.dev> (@0xcharly)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+#ifdef CHARYBDIS_AUTO_POINTER_LAYER_TRIGGER_ENABLE
+#include "timer.h"
+#endif  // CHARYBDIS_AUTO_POINTER_LAYER_TRIGGER_ENABLE
+
+enum charybdis_keymap_layers {
+  LAYER_BASE = 0,
+  LAYER_FUNCTION,
+  LAYER_NAVIGATION,
+  LAYER_MEDIA,
+  LAYER_POINTER,
+  LAYER_NUMERAL,
+  LAYER_SYMBOLS,
+};
+
+// Automatically enable sniping-mode on the pointer layer.
+#define CHARYBDIS_AUTO_SNIPING_ON_LAYER LAYER_POINTER
+
+#ifdef CHARYBDIS_AUTO_POINTER_LAYER_TRIGGER_ENABLE
+static uint16_t auto_pointer_layer_timer = 0;
+
+#ifndef CHARYBDIS_AUTO_POINTER_LAYER_TRIGGER_TIMEOUT_MS
+#define CHARYBDIS_AUTO_POINTER_LAYER_TRIGGER_TIMEOUT_MS 1000
+#endif  // CHARYBDIS_AUTO_POINTER_LAYER_TRIGGER_TIMEOUT_MS
+
+#ifndef CHARYBDIS_AUTO_POINTER_LAYER_TRIGGER_THRESHOLD
+#define CHARYBDIS_AUTO_POINTER_LAYER_TRIGGER_THRESHOLD 8
+#endif  // CHARYBDIS_AUTO_POINTER_LAYER_TRIGGER_THRESHOLD
+#endif  // CHARYBDIS_AUTO_POINTER_LAYER_TRIGGER_ENABLE
+
+enum charybdis_keymap_keycodes {
+#ifdef VIA_ENABLE
+  POINTER_DEFAULT_DPI_FORWARD = USER00,
+#else
+  POINTER_DEFAULT_DPI_FORWARD = SAFE_RANGE,
+#endif  // VIA_ENABLE
+  POINTER_DEFAULT_DPI_REVERSE,
+  POINTER_SNIPING_DPI_FORWARD,
+  POINTER_SNIPING_DPI_REVERSE,
+  SNIPING_MODE,
+  SNIPING_MODE_TOGGLE,
+  DRAGSCROLL_MODE,
+  DRAGSCROLL_MODE_TOGGLE,
+  KEYMAP_SAFE_RANGE,
+};
```

Comment:
A lot of this can, and probably should be moved to the keyboard level.

Reply:
Most of the implementation is at the keyboard level already, available through exposed functions.

IMO, keycodes are better defined at the keymap level than the keyboard.  Otherwise, reconciling custom keycodes definitions in both keyboard and userspace would be tedious, since they both need to start at `USER00`/`SAFE_RANGE`?

Also, while the exposed functionalities at the keyboard level can easily be stripped out of the binary if unused, the same wouldn't be true if the keyboard was defining these keycodes and implementing them, therefore inflating the binary.  We'd need to add more defines and guards to enable/disable features that users actually want to use.

IMO, the current implementation strikes a good balance of "features provided and implemented at the keyboard layer" and "features that users actually want to use in their keymap", but I'm happy to discuss more if you have a different opinion.

Reply:
> IMO, keycodes are better defined at the keymap level than the keyboard. Otherwise, reconciling custom keycodes definitions in both keyboard and userspace would be tedious, since they both need to start at USER00/SAFE_RANGE?

That's the point of including the `KEYMAP_SAFE_RANGE`. 

But also, it's equally tedious re-implementing the keycode functionality on each keymap.  Even more so, that this would have to be implemented separately on both the 3x5 and 4x6 versions.  And I say that, as I am having to re-implement this, in my custom fork.

Reply:
I might be missing something here, so apologies in advance if that's the case.

How do you reconcile both userspace and keyboard implementation defining `KEYMAP_SAFE_RANGE`?

If your userspace starts at `KEYMAP_SAFE_RANGE`, it means that all boards need to define this, even if they do not extend the keycode set.  On the other end, if your userspace starts at `SAFE_RANGE`, and some keyboard extend the keycode set and starts at `SAFE_RANGE` too, then you end up with several keycodes sharing the same value.

Is there some conventions to avoid this scenario, apart from not defining custom keycodes at the keyboard level?

Reply:
What do you think of something like this:

charybdis/rules.mk:
```make
ifeq ($(strip $(CHARYBDIS_CUSTOM_KEYCODES_ENABLE)), yes)
  OPT_DEFS += -DCHARYBDIS_CUSTOM_KEYCODES_ENABLE
endif
```

charybdis/charybdis.h
```c
#ifdef CHARYBDIS_CUSTOM_KEYCODES_ENABLE
enum charybdis_keycodes {
  kc1 = SAFE_RANGE,
  kc2,
  …,
  CHARYBDIS_SAFE_RANGE,
};
#endif // CHARYBDIS_CUSTOM_KEYCODES_ENABLE
```

That way, userspace can use the define:

```c
enum userspace_keycodes {
#ifdef CHARYBDIS_CUSTOM_KEYCODES_ENABLE
  kc1 = CHARYBDIS_SAFE_RANGE,
#elif VIA_ENABLE
  kc1 = USER00,
#else
  kc1 = SAFE_RANGE,
#endif
  kc2,
  …,
  KEYMAP_SAFE_RANGE,
};

And turn `CHARYBDIS_CUSTOM_KEYCODES_ENABLE` off if they want to reduce firmware size.

Reply:
I personally handle it on a per keyboard bases: 
https://github.com/qmk/qmk_firmware/blob/8da9998c9df49e67b4f4679cdbba99d12228fb3a/users/drashna/process_records.h#L20-L24

There isn't a standard way to handle it, aside from having the new safe range listed in a comment in the keymap. 

However, I know that tzarc had some ideas/plans on standardizing the ranges, but that's a while off.

Reply:
I've moved the keycodes definition to the keyboard level.  Users who do not want that can save 310 bytes by defining:

```c
#define NO_CHARYBDIS_KEYCODES
```

Let me know if that works for you.

Reply:
That's perfect!